### ISSUE TITLE
[css-contain] CQ vs shadow boundaries

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#5984]
Can a CQ see past shadow boundaries? If so, exactly how?

If the CQ is done via selectors, the answer is obvious - you can only CQ against elements you can see via selectors. So from within a shadow, the highest ancestor you can CQ against is your host element, no higher.

If CQ is done via an at-rule, the answer is not quite as clear; theoretically, we can walk the box tree (flat tree) to find ancestors. However, we do *not* want a ::part() to be able to see an ancestor within the targeted shadow (it violates shadow encapsulation). And letting a shadow see ancestors outside in the light DOM gives us similar power to :host-context(), which already makes browser vendors unhappy, so maybe we don't want that either. And that's all the cases, so presumably we don't want it to work in either case.

--------

As a related issue, we'll need to specifically define that the CQ pseudo-class matches on host elements, since they're featureless and don't match *anything* besides :host by default otherwise.

**Comment by SebastianZ** [comment]
And with CQ you probably mean container queries, right?

Sebastian

**Comment by svgeesus** [comment]
[probably](https://en.wikipedia.org/wiki/CQ)

Yes.

**Comment by mirisuzanne** [comment]
To clarify, neither approach uses a selector _to target the container_.  Both of these would select a `.card` for conditional styling, and the container remains implicit/contextual:

```css
@container (width > 40em) { .card { ... } }
.card:container(width > 40em) { ... }
```

Maybe they still have different shadow-DOM implications under the hood?

**Comment by mirisuzanne** [comment]
I think I would often want the ability to query the host page - this feature is all about access to contextual information - but that the host element might suffice in most situations (especially for dimensional queries), if we need that limitation.

If I was building a component library with shadow-DOM, I would want content-based components to query external layout containers - and I would likely try to do that with containers named by type (e.g. `@container component (…)` and `@container layout (…)`).

**Comment by bkardell** [comment]
I think a bunch of people are playing with container queries, have we collected any feedback on specifically this from others? It seems like it would be very valuable... Has someone asked for either?

**Comment by lilles** [comment]
The current implementation in Blink uses the flat tree for looking up containers. That was the most straightforward way to implement it. I agree that ::part() and pseudo elements like ::placeholder are problematic if they match containers inside the shadow tree.

This issue should probably be considered together with issue #6711.

**Comment by lilles** [comment]
I think this is an interesting case. Two selectors for the same element in different scopes querying the closest container for inline-size:

```html
<!doctype html>
<style>
  host-element, host-child {
    container-type: inline-size;
  }
  @container (width <= 300px) {
    host-child {}
  }
</style>
<host-element>
  <template shadowroot="open">
    <style>
      inner-container { container-type: inline-size; }
      @container (width <= 200px) {
        ::slotted(host-child) {}
      }
    </style>
    <inner-container>
      <slot></slot>
    </inner-container>
  </template>
  <host-child>Light content</host-child>
</host-element>
```

Do we want host-child to have host-element as its container when the query/selector is in the host-child tree, and have inner-container as the container for the ::slotted rule?

If we just look for the container in the element's own tree, we would not be able to style slotted elements inside the shadow tree. Is that acceptable?

**Comment by lilles** [comment]
With container-name it gets even more interesting.

Should container-name be tree-scoped names?

In the example below, using tree-scoped names would mean that the container rules for host-child below would match differerent containers based on the scope of the rule:

```html
<!doctype html>
<style>
  host-element {
    container-type: inline-size;
    container-name: a;
  }
  @container a (width <= 300px) {
    host-child {}
  }
</style>
<host-element>
  <template shadowroot="open">
    <style>
      inner-container {
        container-type: inline-size;
        container-name: a;
      }
      @container a (width <= 200px) {
        ::slotted(host-child) {}
      }
    </style>
    <inner-container>
      <slot></slot>
    </inner-container>
  </template>
  <host-child>Light content</host-child>
</host-element>
```

**Comment by tabatkins** [comment]
Container names don't need to be tree-scoped; they're not globally registered/visible, but are instead solely looked up via an ancestor search.

We could potentially still *censor* names past a shadow boundary if we wanted, but that would be an independent decision; it's not forced by this issue.

**Comment by mirisuzanne** [comment]
I like the idea of allowing slotted elements to be styled based on the slot location in the shadow tree. It seems likely to me that both styles from inside and outside the shadow boundary are attempting to determine actual space available, and in most cases the shadow parent will give the better answer - but that may not be true in all cases.

**Comment by lilles** [comment]
@tabatkins By ancestor search, do you mean looking up ancestors in the light tree without walking past shadow roots?

If so, it means host-child will never match any containers inside host's shadow tree:

```html
<style>
  @container name (min-width: 100px) { host-child {} }
  @container (min-width: 100px) { host-child {} }
</style>
<host-element><host-child></host-child></host-element>
```

What about slotted rules inside a shadow tree:

```html
<style>
  @container name (min-width: 100px) { ::slotted(div) {} }
  @container (min-width: 100px) { ::slotted(div) {} }
</style>
<div>
  <slot></slot>
</div>
```

Do we start walking from the slot instead of the slotted element to find the container so that it's possible to match containers inside the shadow tree?

**Comment by lilles** [comment]
The current spec covers first case if "descendants" are light tree descendants.

If ::slotted rules should match containers walking from the slot instead of the slotted element being matched, the spec needs to say something about that.

**Comment by andruud** [comment]
> However, we do not want a ::part() to be able to see an ancestor within the targeted shadow (it violates shadow encapsulation).

Possibly stupid question: why is this worse than e.g. `host::part(thing) { background-color: var(--defined-in-shadow-tree); }`?

**Comment by tabatkins** [comment]
> By ancestor search, do you mean looking up ancestors in the light tree without walking past shadow roots?

No, I was referring to jumping up shadow roots as well; CSS *almost always* operates on the flat tree and doesn't care about shadow boundaries, particularly when layout is involved.

That said, while shadows being able to see ancestor light containers seems reasonable, i agree that parts being able to see shadow containers isn't good.

> Possibly stupid question: why is this worse than e.g. host::part(thing) { background-color: var(--defined-in-shadow-tree); }?

Name collisions among variables, while possible, aren't *too* likely in general. On the other hand, the *common* use-case for CQs doesn't even refer to the container by name, so you'll *often* end up accidentally being intercepted by the shadow's container, when you intended to be querying off of some light-dom container that's an ancestor to the shadow.

**Comment by andruud** [comment]
OK, so what about specifying that container selection can't see containers in tree scopes below the scope where the `@container` rule is defined? Basically always flat tree, but skips containers that are below the "originating tree scope".

**Comment by lilles** [comment]
> OK, so what about specifying that container selection can't see containers in tree scopes below the scope where the `@container` rule is defined? Basically always flat tree, but skips containers that are below the "originating tree scope".

That's per definition [shadow-including descendants/ancestors](https://dom.spec.whatwg.org/#concept-shadow-including-ancestor)

**Comment by tabatkins** [comment]
> That's per definition shadow-including descendants/ancestors

No, that's basically flat tree - a ::part(), selecting an element in the shadow, would see shadow ancestors before it escapes back into the light tree and starts seeing light ancestors.

Anders is talking about something more akin to the tree-scoped names rule, where the tree scope of the @container rule affects what container elements are visible to it. A `@container { ::part() {...} }`, since it appears in the light tree's styles, will skip container elements automatically until it's back in the tree scope it was defined in, at which point it can see all ancestor scopes as well.

I think that makes sense?

**Comment by lilles** [comment]
I meant that the part about: "... but skips containers that are below the "originating tree scope" ..." is about shadow-including descendants/ancestors.

**Comment by andruud** [comment]
(After decoding cryptic DOM spec language):

Shadow-including ancestors is almost what we're talking about, but we need to skip the "originating tree scope" based on where the container rule comes from, not based on where the element is in the tree. Otherwise ::slotted will not work as expected.

**Comment by lilles** [comment]
After an off-line discussion with @andruud we think the discussed behavior could be specified as:

In https://drafts.csswg.org/css-contain-3/#query-container, change:

"... styling its descendants ..."

to:

"... styling its shadow-including descendants ..."


And add:

"For selectors with pseudo elements, query containers can be established in the shadow-including inclusive ancestors of the originating element."

Resolving on this could also resolve issue #6711.

1. ::slotted() selectors can query containers inside the shadow tree (including the slot itself)
2. ::part() selectors can query the host, but not internal query containers inside the shadow tree
3. ::placeholder and ::file-selector-button can query the input element, but not any internals if the input element is implemented using Shadow DOM
4. ::before, ::after, ::marker, and ::backdrop queries its originating element
5. ::first-letter and ::first-line queries its originating element, even if the fictional elements may be pushed down past other elements establishing query containers for the purpose of inheritance and rendering.
6. Multiple pseudo elements do not allow pseudo elements to be query containers for other pseudo elements. E.g., the host, but not the part() can be the query container for ::before in: `host::part()::before`. Similarly, the ::before element can not be the query container for the ::marker in: `div::before::marker`.

I have uploaded a Chromium CL with supporting tests for the Shadow DOM cases here:

https://chromium-review.googlesource.com/c/chromium/src/+/3304155

**Comment by andruud** [comment]
:+1: 

But regarding:

> "For selectors with pseudo elements, query containers can be established in the shadow-including inclusive ancestors of the originating element."

we need to describe how to look for containers without mentioning the "kind" of selector.

For e.g. `div::before::marker`, is there a definition which refers to `div` from the perspective of both `::before` and `::marker`? If not, call this "originating element root" or something. Then specify that we look for containers in the shadow-including ancestors starting from the "originating element root".

**Comment by lilles** [comment]
> 👍
> 
> But regarding:
> 
> > "For selectors with pseudo elements, query containers can be established in the shadow-including inclusive ancestors of the originating element."
> 
> we need to describe how to look for containers without mentioning the "kind" of selector.
> 
> For e.g. `div::before::marker`, is there a definition which refers to `div` from the perspective of both `::before` and `::marker`? If not, call this "originating element root" or something. Then specify that we look for containers in the shadow-including ancestors starting from the "originating element root".

I don't think it looks like the spec is written with multiple pseudo elements in mind:

https://drafts.csswg.org/selectors-4/#pseudo-element-attachment

As you say, we either need to clarify that `div` is the originating element for `::marker` in `div::before::marker`, or introduce something like "root originating element", yes.

**Comment by lilles** [comment]
The implementation behind a flag for container queries in Blink now implements the behavior proposed in https://github.com/w3c/csswg-drafts/issues/5984#issuecomment-980694443

Tentative tests landed here: https://wpt.live/css/css-contain/container-queries/container-for-shadow-dom.tentative.html

See https://github.com/w3c/csswg-drafts/issues/6711#issuecomment-1015740083 for tests for pseudo elements.

**Comment by tabatkins** [comment]
Excellent! And I agree that the proposed behavior (use shadow-inclusive descendants/ancestors, with pseudos jumping straight to their originating element) sounds good; it uses only information available at selector-evaluation time, and respects tree scopes reasonably.

**Comment by philkunz** [comment]
Here is example, of why I find the current Spec highly confusing:
https://codepen.io/philkunz/pen/qBVLvWg?editors=1000

container queries break with other logic of styling in dom trees, that we find for example with css variables. Yet container queries serve a similar need: The adaptation of css to context. To have two different context behaviours, does not make sense from my point of view.

**Comment by philkunz** [comment]
@lilles @tabatkins @andruud 
any opinions on the differing behaviours of css variables vs container queries?

**Comment by lilles** [comment]
It was not accidental from my side, and the implementation is in line with the spec.

The rationale for the current spec, is that #child will always match my-component for the container query in the example below. If we use flat tree descendants instead of shadow-including descendants, my-component's shadow tree may intercept the query, either by an implicit size container, if names are not used, or by a name clash (perhaps not likely). That behavior is a bit surprising too as it exposes internals of my-component.

```html
<style>
  @container foo size(width >= 400px) {
    #child { color: green; }
  }
</style>
<my-component style="width:400px; container: foo / inline-size">
  <div id="child">Not green?</div>
</my-component>
```

I don't have very strong opinions here, and I think there are pros and cons to both behaviors.

I think the part about starting from the originating element for selectors with pseudo element still holds, so not exposing shadow internals for ::part() is not affected by the choice.

@tabatkins Was this surprising to you too?

I think what we are looking at here is to possibly change the spec from using "shadow-including descendants" to "flat tree descendants". Changing the implementation in Blink is trivial.

**Comment by philkunz** [comment]
@lilles The thing is: `shadow tree may intercept the query` is the designed way how slotted content should work, I think. 

It should be ENTIRELY up the custom-component how it handles slotted light dom content. It can choose

- not to display it at all. 
- it can choose to iterate over the content with javascript and do something with it (e.g. image carousel)
- it can simply display the slotted content using a <slot>
- it can choose to put the content into a <slot> wrapped by a container

**Comment by lilles** [comment]
> It should be ENTIRELY up to the custom-component how it handles slotted light dom content

True, but custom-component can use container queries with ::slotted() rules inside its shadow dom to style its slotted content based on containers inside the shadow tree, or containers in ancestor trees.

The concern is around the component's internals to affect styling that happens outside of the component. An argument in favor of matching flat tree ancestors, is that the component's users would want to match containers based on the box tree, although the users cannot generally affect what becomes a container of a certain type inside the component, unless the component exposes that possibility via ::part()s.

**Comment by philkunz** [comment]
ok... but isn't the same argument true for css variables? Maybe make an explicit option to not pay respect to shadow tree, otherwise follow what people already know about how context works?

**Comment by lilles** [comment]
> ok... but isn't the same argument true for css variables?

Yes, and for inheritance in general, and that has been controversial in various discussions before.

**Comment by lilles** [comment]
I think the discussion above should be enough for CSSWG members to re-consider whether shadow-including or flat tree ancestor containers could be queried.

If we decide to change, the PROPOSED RESOLUTION would be:

Change the two instances of "shadow-including" in section 4 of css-contain-3 to "flat tree".

**Comment by mirisuzanne** [comment]
Ah, I see what you're saying, @lilles - thanks for expanding on that. 

It does seem to me, from examples like this, that our initial resolution might make sense when thinking of shadow DOM as a 3rd-party widget full of unknowns. But these days, many teams are using custom elements with shadow DOM as the basis for an entire component system – and this resolution makes it impossible for e.g. a card component to query the container in its wrapping layout component. And `::slotted()` isn't much help, because it forces the containing element to understand and apply styles to any elements that get put inside – resulting in a tight coupling of components.

While the initial resolution makes some sense to me, it basically removes the ability for authors to use container queries well, if they are also using shadow DOM slots for their components. I don't think that's a viable tradeoff.

**Comment by tabatkins** [comment]
The spec is indeed as intended right now, and I think it's best to keep it as it is.

As a general rule, CSS features interact with Shadow DOM by allowing the shadow to hide its details as much as possible. This is an intentional part of the design of Shadow DOM; it's intended to be a tool to allow third-party components to be safely used in a page without everyone having to explicitly avoid collisions. This principle guided the development of how @font-face and similar rules are resolved against references to them; how the ::part() pseudo-element works; etc.

The only two big exceptions to this "shadow DOM doesn't expose internal details" principle are layout and inheritance. Layout is pretty much *intrinsically* required to use the flat tree by its very nature, so it does expose whether a slotted element will, say, become a flex item or grid item. Most of the time this doesn't matter very much, tho, as the most important part of a slotted element's layout is its own 'display' value, which it has control of. 

Plain inheritance works over the flat tree, because that's the least likely to break things - for example, if a component is using a particular text and bg color for itself, having slotted elements inherit the text color makes it likely that they'll maintain proper contrast; if they instead inherited their text color from the host element, it would be very easy to have the text become unreadable (for example, if the outer page was using a dark-on-light theme, but the component was styled in a light-on-dark theme).

CQs don't hit these exceptional cases; they're much more similar to @font-face rules. Whether or not a shadow uses a CQ in it shouldn't be exposed to the outer page; if we do expose it, then it becomes part of the component's implied API contract and can't be changed easily. We don't want it to be easy for pages to break when they update a library providing some components.

> It does seem to me, from examples like this, that our initial resolution might make sense when thinking of shadow DOM as a 3rd-party widget full of unknowns. But these days, many teams are using custom elements with shadow DOM as the basis for an entire component system – and this resolution makes it impossible for e.g. a card component to query the container in its wrapping layout component. And ::slotted() isn't much help, because it forces the containing element to understand and apply styles to any elements that get put inside – resulting in a tight coupling of components.
>
> While the initial resolution makes some sense to me, it basically removes the ability for authors to use container queries well, if they are also using shadow DOM slots for their components. I don't think that's a viable tradeoff.

Yes, Shadow DOM can be used in two pretty different ways. If it's purely an organizational tool for first-party content, the restrictions that shadow DOM imposes can be pretty limiting.

However, this is *absolutely* not something that should be adjudicated on a case-by-case basis. Having a "more open" shadow tree is definitely a useful ability, but it should be done at a base level, so that it affects *all* such features at the same time. I am *strongly* against changing CQs to see into shadow trees on an individual basis, as it would make the platform less coherent and more confusing.

**Comment by philkunz** [comment]
@tabatkins Sounds pretty wrong to me.

- You want to break with the logic of css variables (which is the most confusing thing you can do here)
- You want to complicate the API of Custom Components by limiting their ability to act as containers for their light dom children. A custom element should be in control of its light dom children. Thats true for css variables, and it is EXACTLY the same thing for CQs. By limiting the childrens ability to see CQs in a slotted context, you move into global space for handling such things, causing collisions in the aftermath. **Its as wrong as not having scoped element registries from the start, or not enabling font-face in shadow dom... same rubbish.**

What are container queries even good for, if I end up at the browser window after all, when not wanting to manually rewire containers for every slot?

**Comment by tabatkins** [comment]
> Its as wrong as not having scoped element registries from the start, or not enabling font-face in shadow dom... same rubbish.

Yes, not having @font-face work *at all* in shadows was broken, but it wasn't intentional either. Handling it correctly was just more subtle and complex than originally expected. The spec for this *is* now well-defined and correct, and browsers are catching up, and it correctly hides information defined in the shadow.

Like I said, the *intention* of shadow DOM, the reason it was created, was as an info-hiding mechanism to enable third-party components to exist without having to worry about coordinating with the outer page; it was "let's make jQuery UI suck less". (I know, I was there as part of the group defining it.) Shadow DOM is *also* useful for corraling HTML and CSS within a page, but its design is a bit restrictive if used in this way. The solution is not to make individual features less restrictive in an inconsistent manner, but rather to either add a mode to Shadow DOM that is designed for first-party code and generally "transparent" to the outer page, or achieve similar complexity-hiding benefits via new, less restrictive features.

I understand that you're frustrated because you're attempting to use Shadow DOM for first-party content (or at least, third-party content that's designed to explicitly cooperate with the outer page).  That's just a use-case that is, currently, explicitly not handled well.

**Comment by philkunz** [comment]
@tabatkins While I value all your time and contributions deeply, I have like a real anxiety, that there could have been that shiny cool new thing, that could saved thousands of developers valuable lifetime. But instead of it becoming reality now, because of some historic intention what Shadow Dom should have been, someone wants to make me wait another 10 years, until someone else decides, "hey, that could be useful indeed".

**Please, at least give us an option to choose the behaviour!!!**

**Comment by mirisuzanne** [comment]
I totally get the the disconnect between the intention of shadow-DOM and how it has been used. And I agree that tension needs to be addressed at a higher level. It's also pretty clear that the urgency of that conversation is high and increasing. This is not a one-off instance: custom elements have become a popular way to bundle design-system components, and we need to (somehow) address that need in a way that makes more sense.

Setting that aside, I think there is still a difficult tradeoff to consider here - even in relation to third-party code with hidden internals:

> The only two big exceptions to this "shadow DOM doesn't expose internal details" principle are layout and inheritance.

This feature is complicated by the fact that _the purpose of most container queries is to measure layout_ and also sometimes _to query the styles that are being inherited_. That only works if we're measuring the most narrowly relevant context. I'm not convinced that it makes sense to draw this line between 'layout' (which passes through) and 'measuring the layout' (which currently does not). Or between 'property values' (which inherit across the boundary), and 'querying those values' (which we currently disallow).

I also want to acknowledge that there are some pretty decent workarounds here, if we do keep this resolution. Since elements can't query themselves, grid/flex items already require an extra wrapper to act as a measurable container in some cases. Slotted elements would be able to use _the same workaround_ to measure their own available space. With slotted custom elements, the `:host` may often be able to take on that role, without even needing additional elements. It's not an ideal solution in either case, but in practice it seems pretty workable – and a tradeoff we've already had to make for the sake of container queries. 

To summarize: I'm not fully convinced by either the argument that this is essential for maintaining the integrity of Shadow-DOM, or the argument that this is devastating to the entire combined use of these features. It may just be a tradeoff that needs a bit more clarity and discussion.

**Comment by philkunz** [comment]
@mirisuzanne The workaround assumes then, that I want to query the width of the element itself. In many flexible designs though the width of the element is NOT the reference point that you want to query, but another container, that might not correlate in absolute terms to the element.

**Comment by philkunz** [comment]
@mirisuzanne any update on when this is decided?

**Comment by mirisuzanne** [comment]
@philkunz It's on the agenda for the CSSWG, but there's a bit of a backlog. I'm not sure if we'll get to it this week, or maybe next week. Once the conversation happens, the transcript & any decisions will automatically get posted here right away.

**Comment by mirisuzanne** [comment]
@philkunz The WG decided not to revert this, but to start working on a more explicit and general solution to the problem of 'shadow DOM that should be more visible'. (not sure why the transcript/resolution hasn't posted yet). 

In the meantime, the two workarounds are:
- to measure actual slotted space, use a wrapping element
- to measure changes on a less 'immediate' container, the shadow DOM can do the query, and set values on custom props that will inherit into the slot. Depending on the need, those can be values for direct use, or more 'semantic' triggers (like `wide` or `narrow`) that a style query could respond to.

I know that neither of those is ideal, but I think it's true we need a more general fix to this problem. I hope we can make that happen soon.

**Comment by astearns** [comment]
Hand-edited minutes, as the bot dropped before it could post:

RESOLVED: No change, work on more general solution to more openness across shadow boundaries

<details>
<fantasai> Topic: CQ and Shadow Boundaries
<fantasai> github: https://github.com/w3c/csswg-drafts/issues/5984
TabAtkins: Miriam added to agenda
miriam: Issue is, we resolved awhile back that slotted container queries couldn't see containers in the shadow DOM
miriam: which from the perspective of saying that Shadow DOM things should be hidden makes a lot of sense
miriam: Problem is very common use case with web components and shadow DOM these days is to build an entire design system out of them
miriam: you end up with lots of things nested in slots
miriam: and if container queries can't respond, feels limiting
miriam: decent workaround, similar to can't measure grid track with CQ, add an extra element in the slot and measure that
miriam: but wanted to bring that back because feels broken to ppl using the use case
miriam: Proposed path forward is that Container Queries can respond to containers that are in the shadow DOM
miriam: that does expose a bit about what's in the shadow DOM
miriam: but feels similar to exposing inheritance
miriam: we're trying to measure space available, and that's impacted by what's in the shadow DOM
miriam: I can also seem Tab's perspective, there's a disconnect here in the various ways shadow DOM is used
TabAtkins: My position is that containers declared inside a shadow shouldn't be visible to light DOm chilidren distributed in slots
TabAtkins: Shadow DOM was designed to be encapsualted as much as possible
TabAtkins: some purposeful leaks, e.g. ::parts
TabAtkins: and inheritance leaks through the boundaries, but is minor
TabAtkins: but a number of other things we didn't want things to leak
TabAtkins: e.e.g font-faces in shadow can't be used by light DOM
TabAtkins: ...
TabAtkins: if you ever say anything in your styles about that font, won't see that font
TabAtkins: because loaded in Shadow only
TabAtkins: this feels similar
TabAtkins: container in the shadow is not necessarily part of the public contract of the shadow implementation
TabAtkins: might be internal, not to be exposed
TabAtkins: if exposed, then existance of that container *is* exposed
TabAtkins: and becomes part of your contract
TabAtkins: That said, I can see the useful ness of having containers in your shadow, and making that part of your "API" with surrounding content
TabAtkins: This is showing a conflict between shadow DOM as hidden ?? and shadow DOM as organizing mechanism
TabAtkins: ...
TabAtkins: allows cooperating first-party components
TabAtkins: I think we should have a more open shadow DOM that doesn't have info-hiding, and those could expose things like containers and name-defining @rules
TabAtkins: but I don't think that should be an exception to shadow DOM as we have it today
TabAtkins: we should continue to info-hide when possible
TabAtkins: doing this would make it impossible for any UA shadow DOM usage to use containers
TabAtkins: and that's problematic
TabAtkins: so I thin we should keep spec as is, not exposed
miriam: Container in Shadow DOM does impact layout, so that's why it feels weird for it to not impact the measurement
miriam: but I do like the idea of having some explicit control here
TabAtkins: I do get it. Not a good answer right now.
TabAtkins: My preferred decision isn't ideal, it's best given the current tack.
TabAtkins: and we should solve the problem more generally
Rossen_: My current personal lean is more towards trying to find more general solution, that will work for this and others
Rossen_: and not rush into open more dependencies and make shadow DOM less shadowy
<emilio> +1
Rossen_: recognizing the pain as well
Rossen_: Are we in a position to break that tie?
Rossen_: Miriam, how do we feel about this?
miriam: As long as we all feel the pain, and want to work towards a general solution
Rossen_: Miriam, if we resolve to work on more general solution, would it be objectionable?
miriam: No, I would want to work on the more general solution, and teach the workaround for now
Rossen_: sounds like consensus to me, anyone else with a different opinion?
Rossen_: OK, proposed solution is to leave spec as-is and work on a more generalized solution to less shadowy shadow boundaries
Rossen_: any objections?
<fantasai> RESOLVED: No change, work on more general solution to more openness across shadow boundaries
</details>

**Comment by emilio** [comment]
Reopening as per the discussion in #7947. IMO for style queries it seems you really really want the flat tree.

**Comment by lilles** [comment]
I am trying to write up a PR for the spec change, but not entirely sure how to express the pseudo element case.

We cannot keep the "ultimately originating element" as the starting candidate as that would make us incorrectly skip flat tree ancestors for `::slotted()` and `::part()`. But I think we should keep starting from the originating element for `::first-line` and `::first-letter`, and not consider any in-between containers if the originating element is not the inner-most container.

I was thinking about keeping the current wording for pseudo elements which are not [part-like pseudo-elements](https://drafts.csswg.org/css-pseudo-4/#part-like-pseudo-elements), but what about pseudo elements for UA controls like which represent `::file-selector-button`? We should not expose any non-named containers inside that implementation if it uses a UA shadow DOM?

Perhaps it works with just explicitly call out `::slotted()` and `::part()` and change "ultimately originating element" to "originating element" for the other cases?

**Comment by lilles** [comment]
Tests for size and style queries have also been updated. Closing.

**Comment by tabatkins** [comment]
Slightly rephrased how this is defined, tho with no intended behavior change. Now, 'anchor-scope' defines a 'strictly-matched tree-scoped name', and that is defined (in Scoping) to only match against tree-scoped references defined in the exact same tree.
